### PR TITLE
Run tests on Travis CI

### PR DESCRIPTION
**Pull request by abitrolly**
Travis status from my fork https://travis-ci.com/github/abitrolly/cheat.sh/builds

---

They don't really work well.

image

### CODE REVIEW

**Comment by abitrolly** [comment]
@chubin now I am not sure what's going on. More output wouldn't hurt. Like the number of test and the command what is being executed.

https://travis-ci.com/github/abitrolly/cheat.sh/builds/177681447#L425

**Comment by abitrolly** [comment]
It will take a lot of time for me to understand the shell runner. For example, what does this line do?

https://github.com/abitrolly/cheat.sh/blob/c17183cbec3128d9a06c049df1c5791c048463bf/tests/run-tests.sh#L36

**Comment by chubin** [comment]
@abitrolly I believe you are doing it almost right, but it tries to use the standalone mode instead of the curl mode.

Let us take a look at the first test output:

```
CHEATSH_CACHE_TYPE=none python ../lib/standalone.py python/:list
```

Ok, it uses the standalone mode here, and we should curl instead.
If you set:

```
CHEATSH_TEST_STANDALONE=NO
```

What does this mean:

```
1,18d0
< 1line
< :learn
< :list
< Advanced
< Classes
< Comments
< Control_Flow
< Functions
< Modules
< Primitive_Datatypes_and_Operators
< Variables_and_Collections
< doc
< func
< hello
< lambda
< loops
< recursion
< rosetta/
```
?
This is a diff between what you've wanted and what you've got. Because it is running in a standalone mode
which delivers nothing, it is basically the whole expected page.

Regarding this:

```
cat -n tests.txt | sed -n "$(echo "$*" | sed 's/ /p; /g;s/$/p/')"`
```

The line looks so scary but it is pretty trivial:
```
 s{ }{p; }g;
 s{$}{p}
```
what yields:
```
1 2 3 => 1p; 2p; 3p;
```
which then is used as a short script for the outer `sed` printing the selected tests.
I had to document this better; here we just extract the tests that were passed as an arguments
to the `test.sh`.

So, to summarize, `CHEATSH_TEST_STANDALONE=NO` should be sufficient

**Comment by abitrolly** [comment]
@chubin what is the standalone mode? Does it mean ignoring the server?

**Comment by chubin** [comment]
Yes, in the standalone mode you can use it without the server (say during a flight or in a submarine);
currently it does not support programming language queries in this mode though, only cheat sheet queries.
See #14 for some details

**Comment by abitrolly** [comment]
I've discovered the Travis has numerous issues with ANSI escaping. Basically it is doing the same transformation of ANSI sequences as `cheat.sh` does to convert ANSI colored logs to HTML, and it looks like a not very maintainable hack https://github.com/travis-ci/travis-web/blob/master/app/utils/log.js#L236

**Comment by abitrolly** [comment]
Filled a bug https://travis-ci.community/t/ansi-logs-missing-colors-and-formatting-problems/9489

File: .travis.yml [hunk 1]
```diff
@@ -7,3 +7,9 @@ before_install:
   - docker-compose up -d
   - docker ps
   - docker images
+
+script:
+  - sleep 3
```
Comment (SuperSandro2000) [inline .travis.yml:12]: sleep is probably not a good way to know when the server is up. If you run it locally on a slower machine this can easily fail.

File: tests/run-tests.sh [hunk 2]
```diff
@@ -10,8 +10,17 @@
 # 2) configure CHTSH_URL
 # 3) run the script
 
+# work from script's dir
+cd "$(dirname "$0")" || exit
```
Comment (SuperSandro2000) [inline tests/run-tests.sh:14]:
```suggestion
cd "$(dirname "$0")" || exit 1
```

File: tests/run-tests.sh [hunk 3]
```diff
@@ -10,8 +10,17 @@
...
+
+# detect Python - if not set in env, try default virtualenv
 PYTHON="${PYTHON:-../ve/bin/python}"
-"$PYTHON" --version 2>&1 | grep -q 'Python 2' && python_version=2 || python_version=3
+# if no virtalenv, try current python3 binary
+if ! command -v "$PYTHON" &> /dev/null; then
+  PYTHON=$(command -v python3)
```
Comment (SuperSandro2000) [inline tests/run-tests.sh:20]: What if only python3.8 is available? Probably can be ignored.

File: .travis.yml (hunk 1 above)
Comment (abitrolly) [reply .travis.yml:12]: Thanks, cap. So any ideas how to fix that - https://travis-ci.com/github/abitrolly/cheat.sh/builds/177678164#L390? I checked - it is non-trivial

File: tests/run-tests.sh (hunk 2 above)
Comment (abitrolly) [reply tests/run-tests.sh:14]: No. `man exit`
```
       exit [n]
              Cause the shell to exit with a status of n.  If n is omitted, the exit status is that of the last command executed.  A trap on EXIT is executed before
              the shell terminates.
```

File: tests/run-tests.sh (hunk 3 above)
Comment (abitrolly) [reply tests/run-tests.sh:20]: Yes. There is no possible way to check for all Python versions in this simple script.

**Comment by abitrolly** [comment]
@SuperSandro2000 thanks for the comments on things not catched by `shellcheck`. It always good to have more pair of eyes. Any ideas why the tests fail?

File: .travis.yml (hunk 1 above)
Comment (SuperSandro2000) [reply .travis.yml:12]: Retry after one second up to X if the exit code is 52 maybe?

File: tests/run-tests.sh (hunk 2 above)
Comment (SuperSandro2000) [reply tests/run-tests.sh:14]: Funny. I get the reference for exit.h when doing that command.

File: .travis.yml (hunk 1 above)
Comment (abitrolly) [reply .travis.yml:12]: How exit code 52 is related?

**Comment by abitrolly** [comment]
```
redis.exceptions.ConnectionError: Error 99 connecting to localhost:6379. Address not available.

2020-07-31T14:00:00Z {'REMOTE_ADDR': '172.18.0.1', 'REMOTE_PORT': '42306', 'HTTP_HOST': 'localhost:8002', (hidden keys: 20)} failed with ConnectionError
```
At last! Some progress.

**Comment by abitrolly** [comment]
`REDIS_HOST` env setting has no effect. Reading up on host cheat.sh config works https://github.com/chubin/cheat.sh/blob/f5d5fae71218e771e84b896600b2504c0d8814ec/lib/config.py

File: .travis.yml (hunk 1 above)
Comment (SuperSandro2000) [reply .travis.yml:13]:
```
The command "curl http://localhost:8002" exited with 52.
```

Comment (abitrolly) [reply .travis.yml:13]: Then maybe better check until a port is accepting connections?

**Comment by abitrolly** [comment]
@chubin thsi is it. Travis runs test properly. The rest - updating and fixing the tests - could be done in subsequent PRs.

**Comment by abitrolly** [comment]
https://travis-ci.com/github/abitrolly/cheat.sh/builds/178022132

File: tests/run-tests.sh [hunk 4]
```diff
@@ -59,19 +71,24 @@ while read -r number test_line; do
 
   if [ "$test_standalone" = YES ]; then
     test_line="${test_line//cht.sh /}"
-    "${PYTHON}" ../lib/standalone.py "$test_line" > "$TMP" 2> /dev/null
+    [[ $show_details == YES ]] && echo "${PYTHON} ../lib/standalone.py $test_line"
```
Comment (SuperSandro2000) [inline tests/run-tests.sh:74]: You could convert this to a function which takes the string as an argument.

Comment (abitrolly) [reply tests/run-tests.sh:74]: I am not sure I am that bash proficient. Show an example?

Comment (SuperSandro2000) [reply tests/run-tests.sh:74]:
```
log() {
  [[ $show_details == YES ]] && echo "$1" 
}

log("${PYTHON} ../lib/standalone.py $test_line")
```

**Comment by chubin** [comment]
@abitrolly 

The PR is merged, thank you for the great work!

I've fixed the tests, they are almost all green now, except two tests, and they have, probably, the same problem.
Strange enough, this problem occurs only in the test environment, but not when running interactively.

That is how the correct expected behavior looks like:
```
$ curl -s http://localhost:8002/python/copy+file?QT
from shutil import copyfile

copyfile(src, dst)
```

`Q` means *quiet mode*, and `T` means *no terminal*,
but in the test environment the options seems to be ignored.

Could you please take a look at it?
If you will not find anything, then we will investigate deeper

Last build output:

https://travis-ci.org/github/chubin/cheat.sh/builds/715398758

**Comment by abitrolly** [comment]
@chubin I will try to quote them. Maybe there are some problems with shell expansion.

**Comment by chubin** [comment]
@abitrolly I don't think that this helps; try more debug output better instead

**Comment by abitrolly** [comment]
@chubin the part I don't understand is the usage of `eval` - could it be that `eval` does something to `?` like expanding it as a shell pattern?

https://github.com/chubin/cheat.sh/blob/9f99bec1f0293e84d6d8a990a1940c1422e3b0ce/tests/run-tests.sh#L85

**Comment by abitrolly** [comment]
@chubin on my local instance accessing http://localhost:8002/python/copy+file complains that I am offline. Strange that it shows 200 status code. I haven't found a way to debug Flask yet. `logging` doesn't output anything on screen.
```
import logging
logging.error(url)
logging.error(response)
```

**Comment by chubin** [comment]
@abitrolly I am back online; sorry for the disappearance.
I will try to take a look at it and fix the problem

**Comment by abitrolly** [comment]
@chubin no pressure. I will be offline for at least two weeks too. )
